File: Problems 0300-0399/380. Insert Delete GetRandom O(1)/solution.java
class RandomizedSet {
    private List<Integer> nums;
    private Map<Integer, Integer> mp;

    public RandomizedSet() {
        nums = new ArrayList<>();
        mp = new HashMap<>();
    }
    
    public boolean insert(int val) {
        if (mp.containsKey(val)) return false;
        mp.put(val, nums.size());
        nums.add(val);
        return true;
    }
    
    public boolean remove(int val) {
        if (!mp.containsKey(val)) return false;
        
        // We want O(1) time, so remove and replace with last element
        int i = mp.get(val);
        int lastEl = nums.get(nums.size() - 1);
        nums.set(i, lastEl);
        nums.remove(nums.size() - 1);
        mp.put(lastEl, i);
        mp.remove(val);
        return true;
    }
    
    public int getRandom() {
        int n = nums.size();
        int i = (int) (Math.random() * n);
        return nums.get(i);
    }
}

/**
 * Your RandomizedSet object will be instantiated and called as such:
 * RandomizedSet obj = new RandomizedSet();
 * boolean param_1 = obj.insert(val);
 * boolean param_2 = obj.remove(val);
 * int param_3 = obj.getRandom();
 */

File: Problems 0300-0399/380. Insert Delete GetRandom O(1)/solution.py
import random
class RandomizedSet:

    def __init__(self):
        self.nums = []
        self.mp = {}

    def insert(self, val: int) -> bool:
        if val in self.mp:
            return False
        self.mp[val] = len(self.nums)
        self.nums.append(val)
        return True

    def remove(self, val: int) -> bool:
        # Replace num to remove with last number in array
        # since popping off takes O(1)
        if val not in self.mp:
            return False
        numToReplace = self.nums[-1]
        i = self.mp[val]
        self.nums[i] = numToReplace
        self.nums.pop()
        # This line before del line in case
        # of array only has one num
        self.mp[numToReplace] = i
        del self.mp[val]
        return True

    def getRandom(self) -> int:
        # i = random.randint(0, len(self.nums)-1)
        # return self.nums[i]
        return random.choice(self.nums)


# Your RandomizedSet object will be instantiated and called as such:
# obj = RandomizedSet()
# param_1 = obj.insert(val)
# param_2 = obj.remove(val)
# param_3 = obj.getRandom()